KAFKA-16813: Add global timeout for `@ClusterTemplate`, `@ClusterTest` and `@ClusterTests`

m1a2st opened the pull request:
Jira: https://issues.apache.org/jira/browse/KAFKA-16813

add ClusterTest and ClusterTests timout for 30 seconds.

This is testing for GitHub Action CI.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

m1a2st committed:
add timeout 30s annotation to ClusterTest and ClusterTests
and remove timeout for testing with cluster test

chia7712 commented:
@m1a2st please rebase code and fix conflicts

m1a2st committed:
revert line

TaiJuWu commented on core/src/test/java/kafka/test/annotation/ClusterTest.java:
```diff
@@ -34,6 +35,7 @@
 @Target({METHOD})
 @Retention(RUNTIME)
 @TestTemplate
+@Timeout(30)
```
In some case, the timeout is 120 and the other is 30.
Should we set it to 120 to avoid to make test flaky?

WDYT?

m1a2st replied:
Hello @TaiJuWu,
In this PR, we're setting a 30-second timeout limit for all new test infrastructure. We’ve already identified and fixed test cases exceeding this limit in a [previous PR](https://github.com/apache/kafka/pull/16796) and addressed these issues in a follow-up [PR](https://github.com/apache/kafka/pull/16810). Moving forward, developers should ensure their tests complete within this 30-second window to avoid prolonged execution times.
WDYT?

mumrah replied:
Does this work? I know we have tried setting global timeouts in the past with no success.

Also, 30 seconds might be too short for `ClusterTest` tests since these are integration tests. I think 120 is more reasonable.

Can you add a new test that just sleeps for 5 minutes to verify the behavior?

chia7712 replied:
> Does this work? I know we have tried setting global timeouts in the past with no success.

yes, it works. see the QA (https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16957/3/tests) and following error:
```
Error
java.util.concurrent.TimeoutException: testClientInstanceId(kafka.test.ClusterInstance) timed out after 30 seconds
Stacktrace
java.util.concurrent.TimeoutException: testClientInstanceId(kafka.test.ClusterInstance) timed out after 30 seconds
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:272)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	Suppressed: org.apache.kafka.common.errors.InterruptException: java.lang.InterruptedException
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.maybeThrowInterruptException(ConsumerNetworkClient.java:537)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:298)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:253)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:244)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.awaitMetadataUpdate(ConsumerNetworkClient.java:166)
		at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:303)
		at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:264)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.coordinatorUnknownAndUnreadySync(ConsumerCoordinator.java:451)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:1112)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.maybeAutoCommitOffsetsSync(ConsumerCoordinator.java:1144)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.close(ConsumerCoordinator.java:980)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.lambda$close$3(ClassicKafkaConsumer.java:1140)
		at org.apache.kafka.common.utils.Utils.swallow(Utils.java:1050)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1140)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1109)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1097)
		at org.apache.kafka.clients.consumer.KafkaConsumer.close(KafkaConsumer.java:1757)
		at kafka.admin.ClientTelemetryTest.testClientInstanceId(ClientTelemetryTest.java:116)
		at java.lang.reflect.Method.invoke(Method.java:498)
		... 16 more
	Caused by: java.lang.InterruptedException
		... 35 more
```

BTW, it is not a "true" global timeout since it is applied to the tests using `ClusterTest`. That means the tests extending `KafkaServerTestHarness` are not limited.

mumrah replied:
Very nice! 

Is it possible for `@ClusterTest`s to override this annotation for longer timeout?

TaiJuWu replied:
> Hello @TaiJuWu, In this PR, we're setting a 30-second timeout limit for all new test infrastructure. We’ve already identified and fixed test cases exceeding this limit in a [previous PR](https://github.com/apache/kafka/pull/16796) and addressed these issues in a follow-up [PR](https://github.com/apache/kafka/pull/16810). Moving forward, developers should ensure their tests complete within this 30-second window to avoid prolonged execution times. WDYT?

Got it. Sound great. Thanks for your explain.

m1a2st replied:
> Is it possible for `@ClusterTests` to override this annotation for longer timeout?

I don't consider I can achieve this using the `@Timeout` annotation alone. Developers should avoid scenarios where they need to modify this setting. If a test genuinely requires more than 30 seconds, the `@Timeout`annotation can be applied at the method level, it will override this annotation.
```java
    @Timeout(1)
    @ClusterTests({
            @ClusterTest(types = {Type.ZK}),
            @ClusterTest(types = {Type.KRAFT}, serverProperties = {
                    @ClusterConfigProperty(key = AUTHORIZER_CLASS_NAME_CONFIG, value = STANDARD_AUTHORIZER)
            })
    })
    public void testAclCliWithAdminAPI(ClusterInstance cluster) throws InterruptedException {
        Thread.sleep(2000);
        testAclCli(cluster, adminArgs(cluster.bootstrapServers(), Optional.empty()));
    }
```

chia7712 replied:
> Is it possible for @ClusterTests to override this annotation for longer timeout?

yes, it can override the timeout, but each test case created by `@ClusterTests` have isolated elapsed time. That means we don't need to have "a timeout" to include all test cases of a `@ClusterTests`

mumrah commented:
@m1a2st can you pull in the latest trunk changes? I have added code to report the JUnit failures in the Github Actions workflows. I want to do a few runs with this change to ensure we haven't introduced a bunch of timeout failures

chia7712 commented:
@m1a2st could you please add timeout to `@ClusterTemplate`?

m1a2st committed:
add timeout for ClusterTemplate

m1a2st commented:
> @m1a2st  can you pull in the latest trunk changes? I have added code to report the JUnit failures in the Github Actions workflows. I want to do a few runs with this change to ensure we haven't introduced a bunch of timeout failures

I rebased it. 

> @m1a2st could you please add timeout to @ClusterTemplate?

Should I add a constant in TestUtils for 30 second, It will be more easy to controll this parameter.

chia7712 commented:
> I think I should add a constant in TestUtils for 30 second, It will be more easy to controll this parameter. WDYT?

there are only 3 annotations needing the `@Timeout(30)`, so it should be fine to add the timeout annotation individually.

chia7712 commented:
```
| ❌ | LagFetchIntegrationTest | shouldFetchLagsDuringRebalancingWithNoOptimization() | java.lang.AssertionError: 
Expected: <0L>
     but: was <5L> | 63.29s |
```
I have filed https://issues.apache.org/jira/browse/KAFKA-17418 to fix the markdown

m1a2st commented:
I will check these tests which timeout 30 second
- `ResetConsumerGroupOffsetTest.testResetOffsetsToEarliestOnTopicsAndPartitions`
- `ResetConsumerGroupOffsetTest.testResetOffsetsAllTopicsAllGroups`
- `DescribeConsumerGroupTest.testDescribeWithMultiPartitionTopicAndMultipleConsumers`
- `DescribeConsumerGroupTest.testDescribeWithConsumersWithoutAssignedPartitions`
- `ZkMigrationIntegrationTest.testMigrateTopicDeletions`
- `ClientTelemetryTest.testClientInstanceId`
---
This test is timeout for 120s
- `KafkaAdminClientTest.testClientSideTimeoutAfterFailureToReceiveResponse`
---
This is flaky 
- [RestoreIntegrationTest.shouldInvokeUserDefinedGlobalStateRestoreListener](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-16957/5/testReport/junit/org.apache.kafka.streams.integration/RestoreIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldInvokeUserDefinedGlobalStateRestoreListener__/)

chia7712 commented:
`testClientSideTimeoutAfterFailureToReceiveResponse` is known flaky (https://issues.apache.org/jira/browse/KAFKA-15916)

`ZkMigrationIntegrationTest` is a bit complicated, so it should have different timeout. The other IT needs to be improved to reduce the elapsed time. Could you please file sub-tasks (jira) for each test class?

m1a2st commented:
I open these jira tasks to trace these timeout tests
https://issues.apache.org/jira/browse/KAFKA-17471
https://issues.apache.org/jira/browse/KAFKA-17472
https://issues.apache.org/jira/browse/KAFKA-17473

chia7712 commented:
@m1a2st Could you please rebase code to trigger QA again?

chia7712 commented:
@m1a2st Could you please increase the timeout from 30 secs to 60 secs? It seems to me 30 secs is too small. Also, please merge trunk to run CI again

m1a2st committed:
change timeout to 60s

chia7712 commented on core/src/test/java/kafka/test/annotation/ClusterTest.java:
```diff
@@ -34,6 +35,7 @@
 @Target({METHOD})
 @Retention(RUNTIME)
 @TestTemplate
+@Timeout(30)
```
@m1a2st please change it to `30` second :)

m1a2st committed:
transfer to 60s

chia7712 commented:
#17117 is merged now, so @m1a2st could you please rebase code?

m1a2st committed:
Merge branch 'refs/heads/trunk' into gh-KAFKA-17260

m1a2st commented:
Hello @chia7712, 
The fail tests are:
- `OffsetsApiIntegrationTest#testAlterSinkConnectorOffsetsDifferentKafkaClusterTargeted` the reason is can't met condition in 30s
- `PlaintextConsumerSubscriptionTest` the reason is  Failed to close kafka consumer

I think we could keep the 60s timeout for cluster test.

m1a2st commented:
@mumrah, Thanks for your reminder, we only add timeout for cluster tests in this PR, I should modify the title.

chia7712 commented:
@m1a2st Do you forgot to add `ClusterTest` and `ClusterTests` to the title?

m1a2st commented:
I consider the "@TestTemplate" is a symbol for these annotations `ClusterTest`, `ClusterTemplate`, `ClusterTests`, but I will fix the title more clearly.
